plugins/slack/lib/samson_slack/samson_plugin.rb

```diff
@@ -0,0 +1,65 @@
+require 'slack'
+
+module SamsonSlack
+  class Engine < Rails::Engine
+  end
+end
+
+Samson::Hooks.callback :stage_defined do
+  Stage.class_eval do
+    has_many :slack_channels
+    accepts_nested_attributes_for :slack_channels, allow_destroy: true, reject_if: :no_channel_name?
+    validate :channel_exists?
+    before_save :update_channel_id
+
+    def send_slack_notifications?
+      slack_channels.any?
+    end
+
+    def channel_name
+      slack_channels.first.try(:name)
+    end
+
+    def no_channel_name?(slack_attrs)
+      slack_attrs['name'].blank?
+    end
+
+    def update_channel_id
+      if channel_for(channel_name)
```

Comment:
this makes an API call so it should probably be set to a local variable to reuse on the next line. Alternately you could memoise the return value of `channel_for`